src/multi/mod.rs

```diff
@@ -991,3 +998,170 @@ where
     Ok((input, res))
   }
 }
+
+/// Repeats the embedded parser and returns the results in a `Vec`.
+/// Fails if the amount of time the embedded parser is run is not
+/// within the specified range.
+/// # Arguments
+/// * `range` The amount of times to apply the parser. A range without
+/// an upper bound is the same as `(lower..=usize::MAX)`.
+/// * `parse` The parser to apply.
+/// ```rust
+/// # #[macro_use] extern crate nom;
+/// # use nom::{Err, error::ErrorKind, Needed, IResult};
+/// use nom::multi::many;
+/// use nom::bytes::complete::tag;
+///
+/// fn parser(s: &str) -> IResult<&str, Vec<&str>> {
+///   many(0..=2, tag("abc"))(s)
+/// }
+///
+/// assert_eq!(parser("abcabc"), Ok(("", vec!["abc", "abc"])));
+/// assert_eq!(parser("abc123"), Ok(("123", vec!["abc"])));
+/// assert_eq!(parser("123123"), Ok(("123123", vec![])));
+/// assert_eq!(parser(""), Ok(("", vec![])));
+/// assert_eq!(parser("abcabcabc"), Ok(("abc", vec!["abc", "abc"])));
+/// ```
+#[cfg(feature = "alloc")]
+#[cfg_attr(feature = "docsrs", doc(cfg(feature = "alloc")))]
+pub fn many<I, O, E, F, G, H>(
+  range: G,
+  mut parse: F,
+) -> impl FnMut(I) -> IResult<I, Vec<O>, E>
+where
+  I: Clone + InputLength,
+  F: Parser<I, O, E>,
+  E: ParseError<I>,
+  G: IntoRangeBounds<H>,
+  H: RangeBounds<usize>,
+{
+  let range = range.convert();
+  move |mut input: I| {
+    let start = match range.start_bound() {
+      Bound::Included(start) if !range.contains(start) => return Err(Err::Failure(E::from_error_kind(input, ErrorKind::Many))),
+      Bound::Excluded(start) if !range.contains(start) => return Err(Err::Failure(E::from_error_kind(input, ErrorKind::Many))),
+      Bound::Included(start) => Some(*start),
+      Bound::Excluded(start) => Some(*start + 1),
```

Comment:
unclear, why do that, this need an explanation, why use guard here that doesn't help the lecture.

Reply:
It checks for inverted ranges. Previously this was a `min > max` check. It is still the same just expressed using `contains()` instead to handle edge cases with included and unbounded bounds.

Reply:
I know, what I mean that it's not very obvious, also, I still think it's useless check.

Reply:
Outdated.